### DISCUSSION TITLE
Connect to another container using terminal

### DISCUSSION THREAD

**Discussion by Alex-Ferreli** [coder/code-server#2230] [Q&A]
It's possible to connect to another container using terminal?

This is my use case:
2 containers, 1 for a create-react-app, 1 for code-server.

If i call "npm run test" from code-server instance the command is executed inside the "wrong" container, because i do "npm i" only on create-react-app container, so i didn't have these dependencies installed in code-server.

I ended up installing these dependencies too inside code-server container, so i do "npm i" twice.

If it's possible to access create-react-app container from code-server terminal all this issue are covered.

It's possible? Thanks in advance

**Comment by technotablet** [comment]
I run a single container which **also runs the code server**.

An example `docker run` command that I use is

```
docker run -d \
    --restart always \
    --name "create-node-app" \
    -p 8092:8080 \
    -p 8093:3000 \
    -v /Projects/code-server:/code \
    -v /Projects/code-server-configs/default.yml:/code/config.yml \
    -v /Projects/code-server-user-data/vivek:/code/user-data \
    -v /Projects/node-app:/workspace \
    --entrypoint bash \
    node:12 -c "/code/bin/code-server --config /code/config.yml /workspace"
```
* Here my node app runs on port 3000 and is mapped to port 8093 on host machine. You can avoid this and use the ports at runtime too within code-server, but I generally avoid that and define it beforehand.
* The code server is running on port 8080 and is mapped to port 8092. So essentially you can access code-server when you connect to Port 8092.
* I have downloaded the amd64 version of code server on the host machine and I mount it on all containers so that I can share the settings too.

The `default.yml` file contains
```
bind-addr: 0.0.0.0:8080
auth: password
password: mySecretPassword
cert: false
user-data-dir: /code/user-data
```

I put it behind a reverse proxy and a https setup so that it is easier for me to access it on an iPad as well as my primary work machine. You can modify it based on your preferences and environment.

Note: The shared settings work on 3.6.0 (and not 3.6.1).

**Reply by Alex-Ferreli** [reply]
I use the base image of code-server to install custom extensions, i didn't just use the base image. With this approach i lost much of the features i normally use, and i lost in portability too, becouse the machine must have code-server downloaded in some folder.

As much I can understand there isn't a really solution about this, it's impossible to mount a container folder inside another container.

**Reply by technotablet** [reply]
If you consider the way Visual Studio Code works when you are doing remote development - either via SSH or Docker - it installs an instance of a server in the development environment and your "vs code client" connects to that server.

More importantly, the extensions have to be installed separately in that environment ("remote extensions"). Unless you are using a shared storage, each of your containers will have its own set of extensions. If you are using shared storage, then each host machine which has your containers will have those extensions. In any case it's not distributed.

If I understand correctly, code-server is expected to work as the primary interface. So either you can create a workflow which includes code-server as part of the container, or create a package (instead of a docker container) which has all your extensions and settings etc. and save it on a server, and fetch it on each host machine. Thus, instead of a `docker pull`, you'll do a `wget && unzip`  and then restart the containers to use the new code-server version. This would also be portable. See how the great folks at coder.com have implemented [sail.dev](https://sail.dev) environments.

Can you also provide more info on what features you lost that you normally use? If it is just extensions, then you can either use the `user-data-dir` option or https://github.com/cdr/code-server/blob/v3.6.1/doc/FAQ.md#where-are-extensions-stored

**Comment by code-asher** [comment]
One idea is to install the Docker CLI in the code-server container then
mount the Docker socket which would allow communication with the Docker
daemon.

Then you could use `docker exec` from the code-server terminal and get
into the create-react-app container.

If you do this you could also install the Docker plugin which would let
you manage and access all your Docker images from VS Code's UI.

**Reply by technotablet** [reply]
I think if the OP uses `docker exec`, then code-server will just be a better terminal. There would still be a need to run `npm i` twice - one inside the code-server to support the code completion etc., and another inside the create-react-app container to actually run the application.

On the other hand, if he uses the Docker plugin, then there'll be a server instance running inside each container. In that case, the extensions etc. have to be managed separately for each container.

I don't think it is an issue of two containers connecting with each other - it is about how code-server is designed to operate. It is the primary interface, and all code resides inside it for debugging and execution.

**Reply by code-asher** [reply]
True true. My thinking here is that they could use volume mounts to the same directory on the host so both containers would share the same filesystem.

**Reply by code-asher** [reply]
Oh and by the Docker plugin I don't mean the remote Docker plugin, I mean the one that lets you manage the Docker daemon. There wouldn't be any second instance or multiple sets of extensions.

https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker
